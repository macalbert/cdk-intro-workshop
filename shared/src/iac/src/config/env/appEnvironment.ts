export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
