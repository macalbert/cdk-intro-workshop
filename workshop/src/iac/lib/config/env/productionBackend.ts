import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
