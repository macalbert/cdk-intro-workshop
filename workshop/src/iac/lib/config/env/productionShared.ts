import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
