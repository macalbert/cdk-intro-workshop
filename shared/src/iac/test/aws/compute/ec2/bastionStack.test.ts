import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import type { Construct } from "constructs";
import {
	WindowsBastionStack,
	type WindowsBastionStackProps,
} from "../../../../src/aws/compute/ec2/bastionStack";
import { AppEnvironment } from "../../../../src/config/env/appEnvironment";

describe("WindowsBastionStack", () => {
	function createBastion(app: Construct): WindowsBastionStackProps {
		return {
			name: "WindowsBastionStack",
			githubRepo: "shared-test",
			stackName: "shared",
			envName: AppEnvironment.Test,
			vpc: new Vpc(app, "vpc"),
		};
	}

	test("Should_MatchSnapshot_When_WindowsBastionStackIsCreated", () => {
		// Arrange
		const stack = new Stack(undefined, "WindowsBastionStackTest");
		const bastionProps = createBastion(stack);

		// Act
		const bastionStack = new WindowsBastionStack(stack, bastionProps);

		// Assert
		const template = Template.fromStack(bastionStack);
		expect(template.toJSON()).toMatchSnapshot("WindowsBastionStackTest");
	});

	test("Should_CreateResources_When_WindowsBastionStackIsCreated", () => {
		// Arrange
		const stack = new Stack(undefined, "WindowsBastionStackTest");
		const bastionProps = createBastion(stack);

		// Act
		const bastionStack = new WindowsBastionStack(stack, bastionProps);

		// Assert
		const template = Template.fromStack(bastionStack);
		template.resourceCountIs("AWS::EC2::Instance", 1);
		template.resourceCountIs("AWS::EC2::SecurityGroup", 1);
	});
});
