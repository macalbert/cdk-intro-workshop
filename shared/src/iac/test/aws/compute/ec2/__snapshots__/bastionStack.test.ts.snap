// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WindowsBastionStack Should_MatchSnapshot_When_WindowsBastionStackIsCreated: WindowsBastionStackTest 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsserviceamiwindowslatestWindowsServer2022EnglishFullBaseC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/aws/service/ami-windows-latest/Windows_Server-2022-English-Full-Base",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": {
    "Bastion2B4CEC30": {
      "DependsOn": [
        "ec2Role38AB65C0",
      ],
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "IamInstanceProfile": {
          "Ref": "BastionInstanceProfile95B62D1E",
        },
        "ImageId": {
          "Ref": "SsmParameterValueawsserviceamiwindowslatestWindowsServer2022EnglishFullBaseC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.micro",
        "KeyName": {
          "Ref": "KeyPair1622897B",
        },
        "LaunchTemplate": {
          "LaunchTemplateName": "BastionLaunchTemplate",
          "Version": {
            "Fn::GetAtt": [
              "BastionLaunchTemplate47A71220",
              "LatestVersionNumber",
            ],
          },
        },
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "SecurityGroupEc2BastionC8045F3F",
              "GroupId",
            ],
          },
        ],
        "SubnetId": {
          "Fn::ImportValue": "WindowsBastionStackTest:ExportsOutputRefvpcPublicSubnet1Subnet2E65531ECCB85041",
        },
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Name",
            "Value": "shared-test-Test",
          },
          {
            "Key": "Project",
            "Value": "shared-test",
          },
          {
            "Key": "StackId",
            "Value": "m47-shared-test-windowsbastionstack-test-stack",
          },
        ],
        "UserData": {
          "Fn::Base64": "<powershell></powershell>",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "BastionInstanceProfile95B62D1E": {
      "Properties": {
        "Roles": [
          {
            "Ref": "ec2Role38AB65C0",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "BastionLaunchTemplate47A71220": {
      "Properties": {
        "LaunchTemplateData": {
          "MetadataOptions": {
            "HttpTokens": "required",
          },
        },
        "LaunchTemplateName": "BastionLaunchTemplate",
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "KeyPair1622897B": {
      "Properties": {
        "KeyFormat": "pem",
        "KeyName": "keyname-shared-test-test",
        "KeyType": "rsa",
      },
      "Type": "AWS::EC2::KeyPair",
    },
    "SecurityGroupEc2BastionC8045F3F": {
      "Properties": {
        "GroupDescription": "Allow RDP",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "open RDP port",
            "FromPort": 3389,
            "IpProtocol": "tcp",
            "ToPort": 3389,
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "WindowsBastionStackTest:ExportsOutputRefvpcA2121C384D1B3CDE",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ec2Role38AB65C0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
